core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java

```diff
@@ -244,51 +238,42 @@ private void ensureSize() {
 
    private class OffHeapMapSupplier implements Supplier<ConcurrentMap<WrappedBytes,
          InternalCacheEntry<WrappedBytes, WrappedBytes>>> {
-      private final boolean addAllocationSize;
-      private final int allocationSize;
-
-      /**
-       * Supplier that creates a new OffHeapConcurrentMap per invocation
-       * @param addAllocationSize whether starting/stopping map should add the alloction size to count
-       * @param allocationSize rounded up to nearest power of two
-       */
-      private OffHeapMapSupplier(boolean addAllocationSize, int allocationSize) {
-         this.addAllocationSize = addAllocationSize;
-         this.allocationSize = Util.findNextHighestPowerOfTwo(allocationSize);
-      }
-
       @Override
       public ConcurrentMap<WrappedBytes, InternalCacheEntry<WrappedBytes, WrappedBytes>> get() {
-         // OffHeap concurrent map will round allocationSize to nearest power of 2
-         OffHeapConcurrentMap map = new OffHeapConcurrentMap(allocationSize, allocator, offHeapEntryFactory,
-               offHeapListener) {
-            @Override
-            public void stop() {
-               super.stop();
-               if (addAllocationSize) {
-                  lruLock.lock();
-                  try {
-                     currentSize -= UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(allocationSize << 3);
-                  } finally {
-                     lruLock.unlock();
-                  }
-               }
-            }
-         };
-         if (addAllocationSize) {
-            lruLock.lock();
-            try {
-               currentSize += UnpooledOffHeapMemoryAllocator.estimateSizeOverhead(allocationSize << 3);
-            } finally {
-               lruLock.unlock();
-            }
-         }
+         ExpandingOffHeapConcurrentMap map = new ExpandingOffHeapConcurrentMap(allocator, offHeapEntryFactory,
+               offHeapListener);
          map.start();
          return map;
       }
    }
 
-   private class OffHeapListener implements OffHeapConcurrentMap.EntryListener {
+   private class OffHeapListener implements ExpandingOffHeapConcurrentMap.EntryListener {
+
+      @Override
+      public boolean resize(int pointerCount) {
+         if (useCount) {
+            return true;
+         }
+         lruLock.lock();
+         try {
+            long memoryUsed = ((long) pointerCount) << 3;
```

Comment:
```suggestion
            long memoryUsed = ((long) Math.abs(pointerCount)) << 3;
```

`memoryUsed` sounds to me like it should never be negative

Reply:
Okay, I can do that. I would just have to move the negative check up a line as well.